# [5주차 기본과제] 뉴스 기사 분류를 Gemma로 구현하기


In [88]:
import os
from dotenv import load_dotenv

load_dotenv()
HF_TOKEN = os.environ.get("HF_TOKEN")

from huggingface_hub import login

login(HF_TOKEN)
from transformers import AutoTokenizer, AutoModelForCausalLM
import torch

tokenizer = AutoTokenizer.from_pretrained("google/gemma-2b")
model = AutoModelForCausalLM.from_pretrained("google/gemma-2b", device_map="auto")
tokenizer = AutoTokenizer.from_pretrained("google/gemma-2-2b-it")
model = AutoModelForCausalLM.from_pretrained(
    "google/gemma-2-2b-it",
    device_map="auto",
    torch_dtype=torch.bfloat16,
)

The token has not been saved to the git credentials helper. Pass `add_to_git_credential=True` in this function directly or `--add-to-git-credential` if using via `huggingface-cli` if you want to set the git credential as well.
Token is valid (permission: fineGrained).
Your token has been saved to /Users/obov/.cache/huggingface/token
Login successful


Loading checkpoint shards:   0%|          | 0/2 [00:00<?, ?it/s]

tokenizer_config.json:   0%|          | 0.00/47.0k [00:00<?, ?B/s]

tokenizer.model:   0%|          | 0.00/4.24M [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/17.5M [00:00<?, ?B/s]

special_tokens_map.json:   0%|          | 0.00/636 [00:00<?, ?B/s]

Loading checkpoint shards:   0%|          | 0/2 [00:00<?, ?it/s]

In [89]:
from datasets import load_dataset


news_ds = load_dataset("fancyzhx/ag_news")
news_ds

DatasetDict({
    train: Dataset({
        features: ['text', 'label'],
        num_rows: 120000
    })
    test: Dataset({
        features: ['text', 'label'],
        num_rows: 7600
    })
})

In [90]:
world_ds = news_ds["train"].filter(lambda data: data["label"] == 0)
world_ds["text"][:3]

['Venezuelans Vote Early in Referendum on Chavez Rule (Reuters) Reuters - Venezuelans turned out early\\and in large numbers on Sunday to vote in a historic referendum\\that will either remove left-wing President Hugo Chavez from\\office or give him a new mandate to govern for the next two\\years.',
 'S.Koreans Clash with Police on Iraq Troop Dispatch (Reuters) Reuters - South Korean police used water cannon in\\central Seoul Sunday to disperse at least 7,000 protesters\\urging the government to reverse a controversial decision to\\send more troops to Iraq.',
 "Palestinians in Israeli Jails Start Hunger Strike (Reuters) Reuters - Thousands of Palestinian\\prisoners in Israeli jails began a hunger strike for better\\conditions Sunday, but Israel's security minister said he\\didn't care if they starved to death."]

In [91]:
sports_ds = news_ds["train"].filter(lambda data: data["label"] == 1)
sports_ds

Dataset({
    features: ['text', 'label'],
    num_rows: 30000
})

In [92]:
business_ds = news_ds["train"].filter(lambda data: data["label"] == 2)
business_ds

Dataset({
    features: ['text', 'label'],
    num_rows: 30000
})

In [93]:
sci_tech_ds = news_ds["train"].filter(lambda data: data["label"] == 3)
sci_tech_ds

Dataset({
    features: ['text', 'label'],
    num_rows: 30000
})

In [94]:
news_ds["train"].features

{'text': Value(dtype='string', id=None),
 'label': ClassLabel(names=['World', 'Sports', 'Business', 'Sci/Tech'], id=None)}

In [95]:
label_map = news_ds["train"].features["label"].names

In [97]:
news_ds["train"].info

DatasetInfo(description='', citation='', homepage='', license='', features={'text': Value(dtype='string', id=None), 'label': ClassLabel(names=['World', 'Sports', 'Business', 'Sci/Tech'], id=None)}, post_processed=None, supervised_keys=None, builder_name='parquet', dataset_name='ag_news', config_name='default', version=0.0.0, splits={'train': SplitInfo(name='train', num_bytes=29832303, num_examples=120000, shard_lengths=None, dataset_name='ag_news'), 'test': SplitInfo(name='test', num_bytes=1880424, num_examples=7600, shard_lengths=None, dataset_name='ag_news')}, download_checksums={'hf://datasets/fancyzhx/ag_news@eb185aade064a813bc0b7f42de02595523103ca4/data/train-00000-of-00001.parquet': {'num_bytes': 18585438, 'checksum': None}, 'hf://datasets/fancyzhx/ag_news@eb185aade064a813bc0b7f42de02595523103ca4/data/test-00000-of-00001.parquet': {'num_bytes': 1234829, 'checksum': None}}, download_size=19820267, post_processing_size=None, dataset_size=31712727, size_in_bytes=51532994)

In [98]:
from pprint import pprint

pprint(news_ds["train"].info)

DatasetInfo(description='',
            citation='',
            homepage='',
            license='',
            features={'label': ClassLabel(names=['World',
                                                 'Sports',
                                                 'Business',
                                                 'Sci/Tech'],
                                          id=None),
                      'text': Value(dtype='string', id=None)},
            post_processed=None,
            supervised_keys=None,
            builder_name='parquet',
            dataset_name='ag_news',
            config_name='default',
            version=0.0.0,
            splits={'test': SplitInfo(name='test',
                                      num_bytes=1880424,
                                      num_examples=7600,
                                      shard_lengths=None,
                                      dataset_name='ag_news'),
                    'train': SplitInfo(name='train',
   

In [99]:
tokenizer("World", return_tensors="pt")
tokenizer("World", return_tensors="pt")["input_ids"].shape[-1]

2

In [100]:
label_sentences = [
    "Batman: This is mainly about the world, the one we’re sworn to protect.",
    "Joker: Why so serious? Clearly related to Sports news!",
    "Penguin: It’s all about Business, and trust me, I know how to play that game.",
    "Bane: This article mainly covers Sci/Tech.",
]

task_description = """
Task Description:
The scene unfolds with four iconic characters from the Batman universe seated around a table, each reading the same article. Batman, ever composed and logical, is the first to speak, delivering a clear and concise assessment of the piece. The Joker follows, making a mocking and unpredictable remark, turning the topic on its head. Two-Face, as always, relies on the flip of a coin to determine his take. Finally, Bane, with his focus on power and control, offers his perspective, emphasizing the article’s deeper implications.

Your task is simple: “Who explained the article the best?”.

Now, it’s time to choose—who described it best?

ARTICLE:"""

In [101]:
import torch


def zero_shot_classification(
    text, task_description, labels
):  # text는 주어진 입력, task_description은 task에 대한 설명, labels은 class들을 text로 변환한 결과입니다.
    text_ids = tokenizer(task_description + text, return_tensors="pt").to(
        "mps"
    )  # 먼저 task_description과 text를 이어붙인 후, tokenize합니다.

    probs = []

    for label in labels:
        # 그 다음 각 text화된 label들을 tokenize하고 입력에 이어붙인 후, Gemma-2B에 넣어줍니다.
        label_ids = tokenizer(label, return_tensors="pt").to("mps")
        n_label_tokens = label_ids["input_ids"].shape[-1] - 1
        # text로 변환한 label의 token 수를 계산합니다.
        input_ids = {
            "input_ids": torch.concatenate(
                [text_ids["input_ids"], label_ids["input_ids"][:, 1:]], axis=-1
            ),  # concatenate 명령어를 통해 이어붙이는 모습입니다.
            "attention_mask": torch.concatenate(
                [text_ids["attention_mask"], label_ids["attention_mask"][:, 1:]],
                axis=-1,
            ),
        }

        logits = model(**input_ids).logits  # Logit을 계산한 모습입니다.
        prob = 0
        n_total = input_ids["input_ids"].shape[-1]
        for i in range(
            n_label_tokens, 0, -1
        ):  # 일반적으로 text로 변환한 label은 여러 token으로 이루어져있습니다. 이러한 label에 대한 logit은 구성하는 모든 token들의 logit들의 합으로 정의합니다.
            token = label_ids["input_ids"][0, i].item()
            prob += logits[0, n_total - i, token].item()
        probs.append(prob)

        del input_ids
        del logits
        torch.mps.empty_cache()  # 위의 del과 empty_cache() 명령어를 통해 GPU를 제때 할당해제 해줍니다. 만약 GPU가 여유롭다면 지워주시는게 속도적으로 이득입니다.

    return probs

In [102]:
news_ds["train"][1]["text"]

'Carlyle Looks Toward Commercial Aerospace (Reuters) Reuters - Private investment firm Carlyle Group,\\which has a reputation for making well-timed and occasionally\\controversial plays in the defense industry, has quietly placed\\its bets on another part of the market.'

In [103]:
news_ds["train"][1]["label"]

2

In [104]:
text_label = news_ds["train"][179]

probs = zero_shot_classification(
    text_label["text"],
    task_description,
    label_sentences,
)
print(probs)
import numpy as np

print(text_label["label"])
np.argmax(np.array(probs)).item()

[99.28125, 70.71484375, 158.25, 48.431640625]
3


2

In [86]:
def preprocess_function(examples):
    return tokenizer(examples["text"], max_length=200, truncation=True)


tokenized_news_ds = news_ds.map(preprocess_function, batched=True)
tokenized_news_ds

Map:   0%|          | 0/120000 [00:00<?, ? examples/s]

Map:   0%|          | 0/7600 [00:00<?, ? examples/s]

DatasetDict({
    train: Dataset({
        features: ['text', 'label', 'input_ids', 'attention_mask'],
        num_rows: 120000
    })
    test: Dataset({
        features: ['text', 'label', 'input_ids', 'attention_mask'],
        num_rows: 7600
    })
})

In [87]:
import numpy as np
from tqdm import tqdm
from colorama import Fore, Style, Back
import torch.nn.functional as F
import humanize
import math
import random


def calculate_nll(probs_softmax, label):
    return -torch.log(probs_softmax[label])


def get_confidence(probs_softmax):
    top_two_probs, _ = torch.topk(probs_softmax, 2)
    # 가장 높은 확률을 두 번째로 높은 확률로 나누어 비율을 계산
    return math.log10(top_two_probs[0] / top_two_probs[1])


total = 50
n_corrects = 0

# 0부터 7600 사이에서 겹치지 않게 50개의 숫자 뽑기
random_numbers = random.sample(range(7601), total)

for i in tqdm(range(total)):

    print(f"{Fore.YELLOW}----------------------------------{Fore.RESET}")
    text = tokenized_news_ds["test"][random_numbers[i]]["text"]
    label = tokenized_news_ds["test"][random_numbers[i]]["label"]
    probs = zero_shot_classification(
        text,
        task_description,
        label_sentences,
    )

    probs_tensor = torch.tensor(probs)
    probs_softmax = F.softmax(probs_tensor, dim=0)

    pred = np.argmax(np.array(probs))
    if pred == label:
        n_corrects += 1
        tqdm.write(f"Index {i}: {Fore.GREEN}Success 🎉 {Style.RESET_ALL} ")
        print(f"dataset index: {random_numbers[i]}")
        print(f"Confidence: {get_confidence(probs_softmax):.2f}")
        print(f"NLL: {calculate_nll(probs_softmax,label):.2f}")
    else:
        tqdm.write(f"Index {i}: {Fore.RED}Fail 🧐 {Style.RESET_ALL}")
        print(f"dataset index: {random_numbers[i]}")
        print(f"Confidence: {get_confidence(probs_softmax):.2f}")
        print(f"NLL: {calculate_nll(probs_softmax,label):.2f}")
        print(f"Failed Query: {Back.BLACK}{text}{Back.RESET}")
        label, pred = label_map[label], label_map[pred]
        print(
            f"Label : {Fore.CYAN} {label} {Fore.RESET} Pred : {Fore.MAGENTA} {pred} {Fore.RESET}"
        )

print(f"{Fore.RED}>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>{Fore.RESET}")
print(
    f"Total Correct Predictions: {Style.BRIGHT}{n_corrects} ({n_corrects/total:.2%}){Style.RESET_ALL}"
)
print(f"{Fore.RED}>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>{Fore.RESET}")

  0%|          | 0/50 [00:00<?, ?it/s]

----------------------------------


  2%|▏         | 1/50 [00:03<02:55,  3.58s/it]

Index 0: Success 🎉  
dataset index: 6827
Confidence: 25.89
NLL: -0.00
----------------------------------


  4%|▍         | 2/50 [00:15<06:37,  8.28s/it]

Index 1: Fail 🧐 
dataset index: 1089
Confidence: 24.68
NLL: 103.28
Failed Query: Counting the Hops (Forbes.com) Forbes.com - Like Network Appliance, many top tech firms are snapping up Linux programmers, hoping to influence the way the operating system evolves. The trick is to hire programmers closest to Linux creator Linus Torvalds. Torvalds oversees Linux development, but he delegates pieces of the system to the 25 or so code maintainers,like Trond Myklebust at NetApp.Maintainers in turn break their projects into smaller pieces, overseen by submaintainers.
Label :  Sci/Tech  Pred :  Business 
----------------------------------


  6%|▌         | 3/50 [00:20<05:29,  7.02s/it]

Index 2: Fail 🧐 
dataset index: 6472
Confidence: 27.88
NLL: 90.52
Failed Query: Wake Forest survives at the NIT No. 1 Wake Forest used a 19-5 second half run to take the lead and then held off Arizona 63-60 to win the preseason NIT. Mustafa Shakur had a chance to take the lead in the final seconds, but he missed a runner in the lane.
Label :  Sports  Pred :  Business 
----------------------------------


  8%|▊         | 4/50 [00:25<04:43,  6.15s/it]

Index 3: Fail 🧐 
dataset index: 1877
Confidence: 17.70
NLL: 40.76
Failed Query: Sharon acknowledges ignoring road map Prime Minister Ariel Sharon acknowledged that Israel was not following the moribund Mideast peace plan, and said an Israeli pullout from the Gaza Strip was unlikely to revive it, according 
Label :  World  Pred :  Business 
----------------------------------


 10%|█         | 5/50 [00:29<04:06,  5.48s/it]

Index 4: Fail 🧐 
dataset index: 1898
Confidence: 24.11
NLL: inf
Failed Query: AOL Shuns Microsoft Anti-Spam Technology Add America Online Inc. to the growing list of companies and organizations shunning a spam-fighting proposal from Microsoft Corp. AOL cited  quot;tepid support quot; for Microsoft #39;s so-called Sender ID technology, which 
Label :  Sci/Tech  Pred :  Business 
----------------------------------


 12%|█▏        | 6/50 [00:36<04:23,  5.99s/it]

Index 5: Fail 🧐 
dataset index: 3146
Confidence: 30.29
NLL: inf
Failed Query: Judge Challenges EU Position on Microsoft (Reuters) Reuters - A top European Union judge\challenged the EU executive's reasoning in its antitrust court\battle with Microsoft Corp. Friday, questioning why it opposed\the U.S. software giant's setting industry standards.
Label :  Sci/Tech  Pred :  Business 
----------------------------------


 14%|█▍        | 7/50 [00:40<03:43,  5.20s/it]

Index 6: Success 🎉  
dataset index: 2431
Confidence: 26.53
NLL: -0.00
----------------------------------


 16%|█▌        | 8/50 [00:41<02:49,  4.04s/it]

Index 7: Success 🎉  
dataset index: 535
Confidence: 29.26
NLL: -0.00
----------------------------------


 18%|█▊        | 9/50 [00:44<02:30,  3.66s/it]

Index 8: Fail 🧐 
dataset index: 4775
Confidence: 26.87
NLL: inf
Failed Query: NASA Again Postpones Launch of Autonomous DART Spacecraft com Staff. NASA once again postponed the launch of the Demonstration of Autonomous Rendezvous Technology (DART) spacecraft Thursday due to the discovery of contamination inside the fairing of its Pegasus launch vehicle.
Label :  Sci/Tech  Pred :  Business 
----------------------------------


 20%|██        | 10/50 [00:47<02:10,  3.26s/it]

Index 9: Fail 🧐 
dataset index: 4990
Confidence: 25.05
NLL: 57.67
Failed Query: Arafat #39;s brother moved to Cairo for cancer therapy Palestinian sources said on Tuesday that Palestinian leader Yasser Arafat #39;s younger brother Fatehy Arafat was transferred to a hospital in Cairo to be treated for intestines cancer.
Label :  World  Pred :  Business 
----------------------------------


 22%|██▏       | 11/50 [00:48<01:48,  2.77s/it]

Index 10: Fail 🧐 
dataset index: 5095
Confidence: 32.34
NLL: 88.79
Failed Query: Shaq turns up heat in first outing with Miami SHAQUILLE O #39;Neal shot 7-for-9 and finished with 16 points in his Miami debut yesterday as the Heat took a 100-77 victory against the home side New Jersey Nets.
Label :  Sports  Pred :  Business 
----------------------------------


 24%|██▍       | 12/50 [00:50<01:33,  2.47s/it]

Index 11: Fail 🧐 
dataset index: 2171
Confidence: 30.77
NLL: 89.99
Failed Query: Winning NFL Turnover Battle Does Count (AP) AP - Joe Gibbs had seen it before, although he didn't remember: On Dec. 7, 1986, his Washington Redskins turned the ball over seven times and lost to the New York Giants.
Label :  Sports  Pred :  Business 
----------------------------------


 26%|██▌       | 13/50 [00:53<01:42,  2.76s/it]

Index 12: Fail 🧐 
dataset index: 38
Confidence: 31.46
NLL: 72.44
Failed Query: Iran Warns Its Missiles Can Hit Anywhere in Israel  TEHRAN (Reuters) - A senior Iranian military official said  Sunday Israel and the United States would not dare attack Iran  since it could strike back anywhere in Israel with its latest  missiles, news agencies reported.
Label :  World  Pred :  Business 
----------------------------------


 28%|██▊       | 14/50 [00:58<01:55,  3.21s/it]

Index 13: Fail 🧐 
dataset index: 1777
Confidence: 27.25
NLL: 85.73
Failed Query: SportsNetwork Game Preview (Sports Network) - Tim Wakefield tries for his first win in three starts this evening when the Boston Red Sox continue their three-game series with the Tampa Bay Devil Rays at Fenway Park.
Label :  Sports  Pred :  Business 
----------------------------------


 30%|███       | 15/50 [01:01<01:49,  3.14s/it]

Index 14: Fail 🧐 
dataset index: 114
Confidence: 24.81
NLL: inf
Failed Query: Cassini Space Probe Spots Two New Saturn Moons (Reuters) Reuters - Two new moons were spotted around\Saturn by the Cassini space probe, raising the total to 33\moons for the ringed planet, NASA said on Monday.
Label :  Sci/Tech  Pred :  Business 
----------------------------------


 32%|███▏      | 16/50 [01:02<01:31,  2.70s/it]

Index 15: Fail 🧐 
dataset index: 2946
Confidence: 24.68
NLL: 101.49
Failed Query: Large mammals facing the fate of the sabre tooth THE future of the world #39;s large wild mammals is threatened by pressures similar to those that caused the extinction of two-thirds of such species at the end of the most recent Ice Age.
Label :  Sci/Tech  Pred :  Business 
----------------------------------


 34%|███▍      | 17/50 [01:09<02:07,  3.85s/it]

Index 16: Fail 🧐 
dataset index: 2040
Confidence: 27.56
NLL: 85.77
Failed Query: Bonds Takes Stock The distraction of another milestone is gone, and Bonds is preparing himself to play every game for the rest of the season.
Label :  Sports  Pred :  Business 
----------------------------------


 36%|███▌      | 18/50 [01:14<02:13,  4.16s/it]

Index 17: Fail 🧐 
dataset index: 4953
Confidence: 29.21
NLL: 80.32
Failed Query: Redskins lose, Kerry hopes for win TAMPA. Fla. First it was the Boston Red Sox World Series win that had John Kerry grinning, now another sports event has him feeling good.
Label :  Sports  Pred :  Business 
----------------------------------


 38%|███▊      | 19/50 [01:16<01:47,  3.47s/it]

Index 18: Fail 🧐 
dataset index: 4528
Confidence: 26.27
NLL: 60.49
Failed Query: China, ASEAN Agree to End Tariffs (AP) AP - China has reached agreement with the Association of Southeast Asian Nations, or ASEAN, on completely removing tariffs on merchandise goods by 2010 as part of a proposed free trade agreement, the Chinese Ministry of Commerce says.
Label :  World  Pred :  Business 
----------------------------------


 40%|████      | 20/50 [01:19<01:43,  3.43s/it]

Index 19: Success 🎉  
dataset index: 6342
Confidence: 25.06
NLL: -0.00
----------------------------------


 42%|████▏     | 21/50 [01:24<01:50,  3.80s/it]

Index 20: Fail 🧐 
dataset index: 2142
Confidence: 26.59
NLL: inf
Failed Query: TruSecure merges with Betrusted BOSTON - Information security services companies TruSecure Corp. and Betrusted plan to announce on Tuesday that they have merged, forming a new company called Cybertrust.
Label :  Sci/Tech  Pred :  Business 
----------------------------------


 44%|████▍     | 22/50 [01:25<01:28,  3.15s/it]

Index 21: Fail 🧐 
dataset index: 5826
Confidence: 25.20
NLL: inf
Failed Query: AMD readies security, virtualisation features for 2006 Advanced Micro Devices (AMD) plans to build security and virtualisation features into its server processors by 2006, the company said Friday during its annual analyst event.
Label :  Sci/Tech  Pred :  Business 
----------------------------------


 46%|████▌     | 23/50 [01:33<02:05,  4.64s/it]

Index 22: Fail 🧐 
dataset index: 3836
Confidence: 30.94
NLL: 91.30
Failed Query: No. 3 Miami Stops No. 18 Louisville 41-38 (AP) AP - The Louisville Cardinals drew a flag for excessive celebration in the second quarter, and another in the third. Against Miami, the displays of jubilation were premature. Led by Brock Berlin and Devin Hester, the third-ranked Hurricanes erased a 17-point deficit over the final 20 minutes and came from behind twice in the fourth quarter to beat No. 18 Louisville 41-38 Thursday night.
Label :  Sports  Pred :  Business 
----------------------------------


 48%|████▊     | 24/50 [01:35<01:36,  3.71s/it]

Index 23: Success 🎉  
dataset index: 5932
Confidence: 25.57
NLL: -0.00
----------------------------------


 50%|█████     | 25/50 [01:38<01:29,  3.58s/it]

Index 24: Success 🎉  
dataset index: 4572
Confidence: 28.08
NLL: -0.00
----------------------------------


 52%|█████▏    | 26/50 [01:40<01:12,  3.01s/it]

Index 25: Success 🎉  
dataset index: 1678
Confidence: 26.06
NLL: -0.00
----------------------------------


 54%|█████▍    | 27/50 [01:43<01:10,  3.09s/it]

Index 26: Fail 🧐 
dataset index: 6985
Confidence: 27.77
NLL: 63.93
Failed Query: Data revision shows Japan's economy grew slightly in July-September (Canadian Press) Canadian Press - TOKYO (AP) - Japan's economy barely grew during the quarter ending Sept. 30 and in the April-June period it actually shrank instead of squeezing out slight growth, according to revised government data released Wednesday.
Label :  World  Pred :  Business 
----------------------------------


 56%|█████▌    | 28/50 [01:45<01:00,  2.73s/it]

Index 27: Fail 🧐 
dataset index: 4598
Confidence: 26.47
NLL: 60.95
Failed Query: Israeli Officer Arrested Over Killing of Gaza Girl (Reuters) Reuters - Israeli military police on Tuesday\arrested a commander accused by comrades of riddling the body\of a Palestinian schoolgirl with bullets after fellow soldiers\killed her.
Label :  World  Pred :  Business 
----------------------------------


 58%|█████▊    | 29/50 [01:47<00:50,  2.40s/it]

Index 28: Fail 🧐 
dataset index: 5799
Confidence: 26.07
NLL: 60.03
Failed Query: Timing Of Indian Move In Kashmir Vital: Pak Paper Islamabad, Nov. 16 (NNN): Pakistans leading newspaper, Dawn, finds the timing of the Indian announcement on reduction of troops in Kashmir as significance.
Label :  World  Pred :  Business 
----------------------------------


 60%|██████    | 30/50 [01:51<01:02,  3.11s/it]

Index 29: Success 🎉  
dataset index: 6070
Confidence: 26.37
NLL: -0.00
----------------------------------


 62%|██████▏   | 31/50 [01:56<01:09,  3.65s/it]

Index 30: Fail 🧐 
dataset index: 3376
Confidence: 29.12
NLL: 86.71
Failed Query: Mauresmo books a last-eight place FILDERSTADT, Germany -- Amelie Mauresmo stated her determination to stay World No. 1 by surging into the quarterfinals of the Filderstadt Grand Prix in Germany with a 7-5 6-4 win over Patty Schynder.
Label :  Sports  Pred :  Business 
----------------------------------


 64%|██████▍   | 32/50 [02:03<01:20,  4.46s/it]

Index 31: Fail 🧐 
dataset index: 3930
Confidence: 26.42
NLL: 60.83
Failed Query: Sharon to Hold Tense Meeting With Settlers JERUSALEM - After more than a year of avoiding Jewish settlers, Israeli Prime Minister Ariel Sharon has decided to directly confront his former supporters in a meeting about his contentious plan to withdraw from the Gaza Strip and part of the West Bank.    Sharon invited settler leaders to meet with him in Jerusalem on Sunday, just a week before he presents his "disengagement" plan to parliament...
Label :  World  Pred :  Business 
----------------------------------


 66%|██████▌   | 33/50 [02:04<01:02,  3.65s/it]

Index 32: Fail 🧐 
dataset index: 3333
Confidence: 25.08
NLL: 57.76
Failed Query: Blair: Time for Excuses on Africa Is Over (Reuters) Reuters - British Prime Minister\Tony Blair is to say Thursday that the time for excuses on\Africa is over as he chairs a meeting in Ethiopia he hopes will\turn the continent's problems into a global priority.
Label :  World  Pred :  Business 
----------------------------------


 68%|██████▊   | 34/50 [02:08<00:59,  3.74s/it]

Index 33: Fail 🧐 
dataset index: 5953
Confidence: 25.13
NLL: 57.85
Failed Query: Still under fire, US troops shifting to relief effort US forces in Fallujah offer food and first aid as they face scattered pockets of guerrilla resistance.
Label :  World  Pred :  Business 
----------------------------------


 70%|███████   | 35/50 [02:12<00:54,  3.61s/it]

Index 34: Fail 🧐 
dataset index: 2868
Confidence: 27.03
NLL: inf
Failed Query: EBay Expands PayPal Buyer Protection Up to \$1,000 PayPal, eBay Inc. #39;s (EBAY.O: Quote, Profile, Research) online payment service, will expand its US buyer protection program to cover up to \$1,000 for qualified transactions, the company said on Thursday.
Label :  Sci/Tech  Pred :  Business 
----------------------------------


 72%|███████▏  | 36/50 [02:20<01:09,  4.97s/it]

Index 35: Fail 🧐 
dataset index: 2962
Confidence: 29.04
NLL: inf
Failed Query: IBM expands data centers, on-demand service Big Blue enhances its "on demand" offering for companies through its data centers.
Label :  Sci/Tech  Pred :  Business 
----------------------------------


 74%|███████▍  | 37/50 [02:22<00:54,  4.21s/it]

Index 36: Fail 🧐 
dataset index: 4237
Confidence: 27.76
NLL: 87.71
Failed Query: Edmonds homer lifts Cardinals to Game 7 against Astros ST. LOUIS, Missouri All of the wounds in this city, colored a fresh shade of cardinal red, healed with just one touch. The winning pitcher was the one who had the fractured hand.
Label :  Sports  Pred :  Business 
----------------------------------


 76%|███████▌  | 38/50 [02:27<00:51,  4.27s/it]

Index 37: Fail 🧐 
dataset index: 4604
Confidence: 26.94
NLL: inf
Failed Query: Microsoft previews  quot;Whitehorse quot; developer tools Microsoft released on Tuesday a preview version of new tools intended to make it easier for companies to create custom Web applications.
Label :  Sci/Tech  Pred :  Business 
----------------------------------


 78%|███████▊  | 39/50 [02:31<00:48,  4.38s/it]

Index 38: Fail 🧐 
dataset index: 3000
Confidence: 29.98
NLL: 86.94
Failed Query: Sharapova advances to Korea Open final Reigning Wimbledon champion Maria Sharapova has thrashed Anne Kremer of Luxembourg to advance to the final of the Korea Open in Seoul.
Label :  Sports  Pred :  Business 
----------------------------------


 80%|████████  | 40/50 [02:34<00:37,  3.76s/it]

Index 39: Fail 🧐 
dataset index: 2
Confidence: 31.27
NLL: inf
Failed Query: Ky. Company Wins Grant to Study Peptides (AP) AP - A company founded by a chemistry researcher at the University of Louisville won a grant to develop a method of producing better peptides, which are short chains of amino acids, the building blocks of proteins.
Label :  Sci/Tech  Pred :  Business 
----------------------------------


 82%|████████▏ | 41/50 [02:36<00:30,  3.34s/it]

Index 40: Success 🎉  
dataset index: 4824
Confidence: 26.54
NLL: -0.00
----------------------------------


 84%|████████▍ | 42/50 [02:38<00:23,  2.99s/it]

Index 41: Success 🎉  
dataset index: 6000
Confidence: 25.32
NLL: -0.00
----------------------------------


 86%|████████▌ | 43/50 [02:41<00:19,  2.80s/it]

Index 42: Fail 🧐 
dataset index: 7147
Confidence: 25.20
NLL: 103.28
Failed Query: Internet access is free on ferries during tests Tests of high-speed wireless Internet service on one of the region #39;s busiest ferry runs have been canceled, but the service, referred to as Wi-Fi, might still become available on that run, between Seattle and Bremerton, the ferry service said.
Label :  Sci/Tech  Pred :  Business 
----------------------------------


 88%|████████▊ | 44/50 [02:43<00:16,  2.68s/it]

Index 43: Fail 🧐 
dataset index: 4368
Confidence: 30.22
NLL: 92.59
Failed Query: Bucks' Zendon Hamilton Has Knee Surgery (AP) AP - Milwaukee Bucks forward Zendon Hamilton will miss six weeks after undergoing arthroscopic surgery on his right knee.
Label :  Sports  Pred :  Business 
----------------------------------


 90%|█████████ | 45/50 [02:45<00:12,  2.53s/it]

Index 44: Fail 🧐 
dataset index: 3322
Confidence: 29.93
NLL: 94.47
Failed Query: Martnez's Effort Lets Boston Fans Take Heart Boston's Pedro Martnez put his past four starts behind him Wednesday night to post a redemptive 8-3 victory over the Anaheim Angels.
Label :  Sports  Pred :  Business 
----------------------------------


 92%|█████████▏| 46/50 [02:47<00:09,  2.39s/it]

Index 45: Success 🎉  
dataset index: 1188
Confidence: 25.08
NLL: -0.00
----------------------------------


 94%|█████████▍| 47/50 [02:49<00:07,  2.36s/it]

Index 46: Fail 🧐 
dataset index: 3947
Confidence: 22.45
NLL: 84.88
Failed Query: Testing didn #39;t curtail homers It #39;s time to fess up. We were among the hordes of skeptics (sheep?) who boldly proclaimed drug-testing would blow a hole in the number of runs and home runs we #39;d see in 2004.
Label :  Sports  Pred :  Business 
----------------------------------


 96%|█████████▌| 48/50 [02:59<00:09,  4.54s/it]

Index 47: Fail 🧐 
dataset index: 4268
Confidence: 29.39
NLL: inf
Failed Query: Strong Server, PC Sales Boost Microsoft Revenue The company's earnings beat Wall Street expectations.
Label :  Sci/Tech  Pred :  Business 
----------------------------------


 98%|█████████▊| 49/50 [03:01<00:03,  3.87s/it]

Index 48: Fail 🧐 
dataset index: 5327
Confidence: 21.78
NLL: 77.19
Failed Query: OU puts exclamation point on season No apologies, no justifications, no excuses necessary. Oklahoma isn #39;t slinking into the BCS championship game through the back door this year.
Label :  Sports  Pred :  Business 
----------------------------------


100%|██████████| 50/50 [03:04<00:00,  3.68s/it]

Index 49: Fail 🧐 
dataset index: 3772
Confidence: 25.43
NLL: 81.20
Failed Query: US moves on to final round The United States national soccer team revealed both its immediate and long-term future in a 6-0 victory over Panama last night.
Label :  Sports  Pred :  Business 
>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>
Total Correct Predictions: 11 (22.00%)
>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>


In [32]:
# for data in news_ds["test"]:
#     print(data)
[data["text"] for data in news_ds["test"]][26:50]

['Giddy Phelps Touches Gold for First Time Michael Phelps won the gold medal in the 400 individual medley and set a world record in a time of 4 minutes 8.26 seconds.',
 "Tougher rules won't soften Law's game FOXBOROUGH -- Looking at his ridiculously developed upper body, with huge biceps and hardly an ounce of fat, it's easy to see why Ty Law, arguably the best cornerback in football, chooses physical play over finesse. That's not to imply that he's lacking a finesse component, because he can shut down his side of the field much as Deion Sanders ...",
 "Shoppach doesn't appear ready to hit the next level With the weeks dwindling until Jason Varitek enters free agency, the Red Sox continue to carefully monitor Kelly Shoppach , their catcher of the future, in his climb toward the majors. The Sox like most of what they have seen at Triple A Pawtucket from Shoppach, though it remains highly uncertain whether he can make the adjustments at the plate ...",
 "Mighty Ortiz makes sure Sox can r

In [34]:
news_ds["test"][0]

{'text': "Fears for T N pension after talks Unions representing workers at Turner   Newall say they are 'disappointed' after talks with stricken parent firm Federal Mogul.",
 'label': 2}

---


# 실습: Zero-shot Classification

이번 실습에서는 open LLM을 가지고 zero-shot classification을 해봅니다. 먼저 필요한 library들을 설치합시다.


그 다음 Gemma-2B를 사용하기 위해 다음과 같은 작업을 진행합니다:

1. huggingface.co 계정 만들고 로그인하기
2. https://www.kaggle.com/models/google/gemma/license/consent 에서 Gemma license 동의하기
3. 홈 화면으로 돌아와, `Profile > Settings > Access Tokens` 메뉴로 들어와 "Write" type의 token 생성하기
4. 생성한 토큰을 아래 "HF TOKEN"에 불여넣고 셀을 실행하기.


In [ ]:
import os
from dotenv import load_dotenv

load_dotenv()
HF_TOKEN = os.environ.get("HF_TOKEN")

from huggingface_hub import login

login(HF_TOKEN)

In [ ]:
from transformers import AutoTokenizer, AutoModelForCausalLM

tokenizer = AutoTokenizer.from_pretrained("google/gemma-2b")
model = AutoModelForCausalLM.from_pretrained("google/gemma-2b", device_map="auto")

이번에는 Gemma-2B를 가지고 간단한 text 생성을 해봅시다.
"What is your name?" 이라는 text를 넣었을 때 어떤 text가 생성되는지 살펴봅시다.


In [ ]:
input_text = "What is your name?"
input_ids = tokenizer(input_text, return_tensors="pt").to("mps")
input_ids

In [ ]:
input_text = "What is your name?"
input_ids = tokenizer(input_text, return_tensors="pt").to("mps")
outputs = model.generate(**input_ids)
print(tokenizer.decode(outputs[0]))

In [ ]:
tokens = input_ids["input_ids"]
print(tokens)
input_ids

2B의 작은 LLM이라 질좋은 답변이 나오지 않는 것을 알 수 있습니다.
이번에는 입력으로 넣어준 token들의 logit을 계산해봅시다.


In [ ]:
tokens = input_ids["input_ids"]
print(tokens)

logits = model(**input_ids).logits
for i in range(tokens.shape[-1]):
    token = tokens[0, i].item()
    print(logits[0, i, token])

위와 같이 모델 출력의 `.logits`을 통해 token들의 logit을 알 수 있습니다.
Logit은 높을 수록 token이 나올 확률이 높다는 뜻입니다.

이번에는 logit 계산을 통해 zero-shot classification을 구현해보도록 하겠습니다.


In [8]:
import torch


def zero_shot_classification(
    text, task_description, labels
):  # text는 주어진 입력, task_description은 task에 대한 설명, labels은 class들을 text로 변환한 결과입니다.
    text_ids = tokenizer(task_description + text, return_tensors="pt").to(
        "mps"
    )  # 먼저 task_description과 text를 이어붙인 후, tokenize합니다.

    probs = []

    for label in labels:
        # 그 다음 각 text화된 label들을 tokenize하고 입력에 이어붙인 후, Gemma-2B에 넣어줍니다.
        label_ids = tokenizer(label, return_tensors="pt").to("mps")
        n_label_tokens = label_ids["input_ids"].shape[-1] - 1
        # text로 변환한 label의 token 수를 계산합니다.
        input_ids = {
            "input_ids": torch.concatenate(
                [text_ids["input_ids"], label_ids["input_ids"][:, 1:]], axis=-1
            ),  # concatenate 명령어를 통해 이어붙이는 모습입니다.
            "attention_mask": torch.concatenate(
                [text_ids["attention_mask"], label_ids["attention_mask"][:, 1:]],
                axis=-1,
            ),
        }

        logits = model(**input_ids).logits  # Logit을 계산한 모습입니다.
        prob = 0
        n_total = input_ids["input_ids"].shape[-1]
        for i in range(
            n_label_tokens, 0, -1
        ):  # 일반적으로 text로 변환한 label은 여러 token으로 이루어져있습니다. 이러한 label에 대한 logit은 구성하는 모든 token들의 logit들의 합으로 정의합니다.
            token = label_ids["input_ids"][0, i].item()
            prob += logits[0, n_total - i, token].item()
        probs.append(prob)

        del input_ids
        del logits
        torch.cuda.empty_cache()  # 위의 del과 empty_cache() 명령어를 통해 GPU를 제때 할당해제 해줍니다. 만약 GPU가 여유롭다면 지워주시는게 속도적으로 이득입니다.

    return probs

아래는 실제로 zero-shot classification을 해본 결과입니다.


In [ ]:
probs = zero_shot_classification(
    "I am happy!", "Is the sentence positive or negative?: ", ["positive", "negative"]
)
print(probs)

보시다시피 우리는 Gemma를 별도로 학습하지 않았음에도 불구하고 주어진 문장이 긍정적이라는 것을 정확하게 예측하고 있습니다.

다음은 영화 리뷰 감정 분석 task에 적용해봅시다.
먼저 data를 불러옵니다.


In [ ]:
from datasets import load_dataset


imdb = load_dataset("imdb")


def preprocess_function(examples):
    return tokenizer(examples["text"], max_length=200, truncation=True)


tokenized_imdb = imdb.map(preprocess_function, batched=True)

그리고 `test` data에서 50개의 영화 리뷰에 대해 예측하는 코드는 다음과 같습니다.


In [ ]:
import numpy as np
from tqdm import tqdm


n_corrects = 0
for i in tqdm(range(50)):
    text = tokenized_imdb["test"][i]["text"]
    label = tokenized_imdb["test"][i]["label"]
    probs = zero_shot_classification(
        text,
        "A movie review is given. Decide that the movie review is positive or negative: ",
        labels=["Answer: negative.", "Answer: positive."],
    )

    pred = np.argmax(np.array(probs))
    if pred == label:
        n_corrects += 1

print(n_corrects)

In [ ]:
text = tokenized_imdb["test"][i]["text"]
text

In [ ]:
label = tokenized_imdb["test"][i]["label"]
label

보시다시피 정확도 88%로, 매우 높은 성능을 보이는 것을 알 수 있습니다.
